os/os-mac.h

```diff
@@ -117,3 +123,58 @@ static inline bool os_cpu_has(cpu_features feature)
 }
 
 #endif
+
+/*
+ * discard_pages should not be run within Rosetta. Native arm64 or x86 only. 
+ */
+
+#define MMAP_CHUNK_SIZE		(1024 * 1024 * 1024)
+
+static inline int discard_pages(int fd, off_t offset, off_t size)
+{
+	caddr_t *addr;
+	uint64_t chunk_size = MMAP_CHUNK_SIZE;
+
+	if (fsync(fd) < 0) {
+                int __err = errno; 
+		log_err("%s: Cannot fsync file: %s \n", __func__, strerror(errno));
+                errno = __err;
+		return -1;
+	}	    
+    
+	/*
+	 * mmap the file in 1GB chunks and msync(MS_INVALIDATE).
+	 */
+	while (size > 0) {
+		uint64_t mmap_size = MIN(chunk_size, size);
+
+		addr = mmap((caddr_t)0, mmap_size, PROT_NONE, MAP_SHARED, fd, offset);
+		if (addr == MAP_FAILED) {
+                        int __map_errno = errno;
+			fprintf(stderr, "Failed to mmap (%s), offset = %llu, size = %llu\n",
```

Comment:
@Developer-Ecosystem-Engineering I've just looked in `printf(3)` and it makes no mention of `errno` so I wonder why you have to save and restore it...

@axboe do we think that it's possible that printing the error message could end up overwriting `errno`?